packages/@o3r/test-helpers/src/jest-environment/index.ts

```diff
@@ -0,0 +1,118 @@
+import type { EnvironmentContext, JestEnvironmentConfig } from '@jest/environment';
+import type { Circus } from '@jest/types';
+import { TestEnvironment as NodeTestEnvironment } from 'jest-environment-node';
+import { execSync } from 'node:child_process';
+import pidFromPort from 'pid-from-port';
+import { rm } from 'node:fs/promises';
+import path from 'node:path';
+import { prepareTestEnv, type PrepareTestEnvType } from '../prepare-test-env';
+
+export type TestEnvironment = Awaited<ReturnType<typeof prepareTestEnv>>;
+declare global {
+  // eslint-disable-next-line no-var
+  var testEnvironment: TestEnvironment;
+}
+
+const rootFolder = path.join(__dirname, '..', '..', '..', '..');
+
+/**
+ * Custom Jest environment used to manage test environments with Verdaccio setup
+ */
+export class JestEnvironmentO3r extends NodeTestEnvironment {
+  /**
+   * Stop Verdaccio server at the end of the test
+   * @private
+   */
+  private shouldHandleVerdaccio = false;
+
+  /**
+   * Folder containing the generated files
+   * @private
+   */
+  private readonly appFolder: string;
+
+  /**
+   * Increment used in folder name in case of multiples tests
+   * @private
+   */
+  private appIndex = 0;
+
+  /**
+   * Type of test environment to be created
+   * @private
+   */
+  private readonly prepareTestEnvType: PrepareTestEnvType | undefined;
+
+  /**
+   * Map test name with test environment
+   * @private
+   */
+  private testEnvironments: Record<string, TestEnvironment> = {};
+
+  constructor(config: JestEnvironmentConfig, context: EnvironmentContext) {
+    super(config, context);
+    this.appFolder = context.docblockPragmas['jest-environment-o3r-app-folder'] as string;
+    this.prepareTestEnvType = context.docblockPragmas['jest-environment-o3r-type'] as PrepareTestEnvType | undefined;
+  }
+
+  /**
+   * Start Verdaccio server if not already running on localhost:4873
+   * @private
+   */
+  private async startVerdaccio() {
+    try {
+      await pidFromPort(4873);
+    } catch (ex) {
+      this.shouldHandleVerdaccio = true;
+      execSync('yarn verdaccio:start', {cwd: rootFolder, stdio: 'inherit'});
+      execSync('yarn verdaccio:publish', {cwd: rootFolder, stdio: 'inherit'});
+    }
+  }
+
+  /**
+   * Stop Verdaccio server if it was started by this class
+   * @private
+   */
+  private stopVerdaccio() {
+    if (this.shouldHandleVerdaccio) {
+      execSync('yarn verdaccio:stop', {cwd: rootFolder, stdio: 'inherit'});
+    }
+  }
+
+  /**
+   * Catch Jest lifecycle events
+   * @param event
+   * @param _state
+   */
+  public async handleTestEvent(event: Circus.AsyncEvent, _state: Circus.State) {
+    // Create test environment before test starts
+    if (event.name === 'test_start') {
+      const appFolder = `${this.appFolder}${this.appIndex++ || ''}`;
```

Comment:
Is the use of the increment in a function always evaluated before its use in a function for node?
Is this a reliable behaviour?

Reply:
`++a` returns the value after increment
`a++` returns the value before increment